### ISSUE TITLE
Spec does not specify if and how multiple registrations for the same method will be handled.

### ISSUE THREAD

**Issue by flyx** [microsoft/language-server-protocol#1069]
In **Register Capability**, the spec reads

> Server must not register the same capability both statically through the initialize result and dynamically for the same document selector.

(It would be nice to clarify what happens if no *document selector* is available but that is not my issue here).

Now this implies that it is okay to register for the same capability if you use different document selectors. However a `DocumentSelector` is a list of `DocumentFilter` so ever if two document selectors are different, they may still contain the same filter, or different filters that can still match the same document. Also, multiple dynamic registrations seem to be okay even with the same document filter.

So, what is the semantic here? Will a new registration for a method *m* replace a previous registration for *m*? Will both registrations pile up? If so, will two requests be sent when a document matches the selectors of multiple registrations? How can the server know which registration caused the request to be sent?

I think there are two possibilities to answer these questions:

 * Either the spec defines that only one registration can be active for each method. Then the server always knows which selector triggered a request.
 * Or the spec defines explicitly that multiple registrations are okay for each method. Then, the server should have a possiblity to check which selector triggered a notification (either by sending a list of registration ids in one request, or by sending multiple requests with a different registration id each). If the client does not provide this option, the server would need to replicate the document selector filtering which does seem superfluous if the client already did it.

In any case, the spec should be clear about what happens in this case. Currently, I can't find any information about it.

**Comment by dbaeumer** [comment]
I agree and this came up in internal discussions as well. Our current thinking is to allow more than one registration and provide and indication which registration trigger the request.

**Comment by flyx** [comment]
I am currently implementing a server. My impression is that a good way to handle this would be:

 * Let the server send some ID with registration requests. Client requests will send a list of associated registration IDs if any.
 * The IDs only need to be unique in the context of one method. Different methods may take the same ID for a registration. 
 * The server can implement multiple modules that are each identified by such an ID (e.g. for different languages). Messages can be dispatched to the correct module via ID.
 * By only defining the ID and not its semantic in the protocol, it's flexible enough to support different use cases.
 * *maybe* give registrations a priority so that if multiple registrations match, the one with the highest priority gets to supply its ID to the message.

**Comment by reflectronic** [comment]
The behavior of `vscode-languageserver-node`, generally, is to send duplicate messages to the server for each registration. For example, if a server registers `textDocument/didChange` twice, VS Code sends _two_ notifications to the server each time a document is changed. (This happens because, each time a server sends a registration for e.g. `textDocument/didChange`, VS Code adds another set of "text changed" event handlers that result in notifications being sent.)

This behavior isn't quite uniform across all methods. (For example, VS Code duplicates _incremental_ `textDocument/didChange` notifications, but never sends more than one _full_ `textDocument/didChange` notification.) Peeking at the code suggests that the current behavior is largely a result of implementation details rather than deliberate design.

Despite this, some language servers in the ecosystem are already starting to expect VS Code's behavior:

> For example, the C# language server uses multiple registrations for `textDocument/diagnostic`. This allows it to use one request to compute only syntax errors (which is fast), another to compute semantic errors (which is slower), a third to run code analysis plugins, etc. This way, syntax errors do not have to wait for potentially slow code analyzers to complete before being returned to the user. 

Given that servers are already using these semantics, they should be precisely specified for the benefit of the ecosystem. It is very tedious, as a client implementer, to verify exactly how `vscode-languageserver-node` handles duplicate registrations for every method.

**Comment by reflectronic** [comment]
@dbaeumer I'd like to get your thoughts on this :-)

**Comment by angelozerr** [comment]
> Or the spec defines explicitly that multiple registrations are okay for each method

I also implemented this feature for IntelliJ by studying the behavior of vscode with multiple languzge servers like gopls, metalls, rust-analyzer and the vscode behavior seems following this rule.

**Comment by dbaeumer** [comment]
I fully agree that this needs a better specification. My current thinking was that a server can still register more than on handler but will have an indication which registration is called. This is currently not possible. Such an implementation can even be made backwards compatible.
